### PR TITLE
pidstore: do not delete all external pids when it's unecessary

### PR DESCRIPTION
**Pull request by pazembrz**
* Additionally fix cnum provider when creating cnum from data provided.
 * Fix ORCID generator in tests

 * INSPIR-3641

### CODE REVIEW

File: backend/inspirehep/pidstore/minters/base.py [hunk 1]
```diff
@@ -78,11 +79,42 @@ def mint(cls, object_uuid, data):
...
+                object_type=minter.object_type,
+            )
+            .filter(PersistentIdentifier.status != PIDStatus.DELETED)
+            .with_entities("pid_value")
+            .all()
+        }
+        pids_requested = minter.get_pid_values()
+        pids_to_delete = pids_in_db - pids_requested
+        pids_to_create = pids_requested - pids_in_db
+        minter.delete(object_uuid, None, pids_to_delete)
```
Comment (drjova) [inline backend/inspirehep/pidstore/minters/base.py:98]: If we don't need to pass `data` anymore let's change the signature of delete, it's used only internaly.

File: backend/inspirehep/pidstore/minters/doi.py [hunk 2]
```diff
@@ -18,4 +18,4 @@ def create(self, pid_value):
 
     def get_pid_values(self):
         pid_values = {str(pid_value).lower() for pid_value in super().get_pid_values()}
-        return pid_values
+        return set(pid_values)
```
Comment (drjova) [inline backend/inspirehep/pidstore/minters/doi.py:21]: isn't already a set?

Comment (pazembrz) [reply backend/inspirehep/pidstore/minters/doi.py:21]: :+1:  Yes, you are right. Here it's already set.

File: backend/inspirehep/pidstore/minters/base.py (hunk 1 above)
Comment (pazembrz) [reply backend/inspirehep/pidstore/minters/base.py:98]: we still use data in `delete` methods in `ControlNumberMinter` and `CNUMMinter`

File: backend/tests/integration/pidstore/minters/test_minters_arxiv.py [hunk 3]
```diff
@@ -110,3 +111,50 @@ def test_mitner_arxiv_eprints_missing_schema(inspire_app):
 
     with pytest.raises(MissingSchema):
         ArxivMinter.mint(record_id, record_data)
+
+
+def test_minter_proper_pid_management(inspire_app):
```
Comment (drjova) [inline backend/tests/integration/pidstore/minters/test_minters_arxiv.py:116]: could we add a test with more pid types? a combination of internal and external?

Comment (pazembrz) [reply backend/tests/integration/pidstore/minters/test_minters_arxiv.py:116]: All tests for externals are actually tests with internal also (control_number is internal) But I will add few more tests.

File: backend/tests/helpers/providers/record_provider.py [hunk 4]
```diff
@@ -39,12 +39,23 @@ def arxiv():
...
+        val = str(orcid_id).replace("-", "").replace(" ", "").upper()
+        r = 0
+        for x in val:
+            r = (r + int(x)) * 2
+        ck = (12 - r % 11) % 11
+        if ck == 10:
+            ck = "X"
+
+        orcid_id += str(ck)
+        return orcid_id
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply backend/tests/helpers/providers/record_provider.py:57]: This is calculation of checksum which is integral part of ORCID, and it's checked, so without it generated ORCID will be wrong and will result in record creation failure if used with `create_record` helper.

File: backend/tests/integration/pidstore/minters/test_minters_arxiv.py [hunk 5]
```diff
@@ -110,3 +111,50 @@ def test_mitner_arxiv_eprints_missing_schema(inspire_app):
 
     with pytest.raises(MissingSchema):
         ArxivMinter.mint(record_id, record_data)
+
+
+def test_arxiv_minter_proper_pid_management(inspire_app):
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply backend/tests/integration/pidstore/minters/test_minters_arxiv.py]: :+1:

File: backend/tests/integration/pidstore/minters/test_minters_control_number.py [hunk 6]
```diff
@@ -326,3 +325,24 @@ def test_control_number_seminars_without_control_number(inspire_app):
...
+
+def test_control_number_minter_proper_pid_management(inspire_app):
+    record = create_record("lit")
+    expected_cn = str(record["control_number"])
+    cn_pid = PersistentIdentifier.query.filter_by(pid_type="lit").one()
+
+    assert cn_pid.pid_value == expected_cn
+
+    data = dict(record)
+    data["control_number"] = f"{expected_cn}1"
```
Comment (drjova) [inline backend/tests/integration/pidstore/minters/test_minters_control_number.py:338]: there is `1` at the end, is this correct?

Comment (pazembrz) [reply backend/tests/integration/pidstore/minters/test_minters_control_number.py:338]: Yes, I Intentionally added 1 to change control_number to be sure that it's different to test behavior when CN changes.

**Pull request merged**